### PR TITLE
Allow embed sandbox to make content non-interactive for unknown embeds

### LINKED ISSUE #7872: Embed Block: giphy embeds switch from animated gif to html page when clicked
**Describe the bug**
Within Gutenberg, if you add a giphy embed via the generic embed block, the properly gif will show but if you then click on the gif to adjust the layout, the image showing changes to the giphy homepage. 

**To Reproduce**
1. Create a new post.
2. Using the embed block, add a giphy link.
3. See the gif embed
4. Click on the embed to adjust the layout of the embed. 
5. Watch the gif change instead.
6. Preview the post, and you will see the correct embed. 

**Expected behavior**
I expected the same preview image to show consistently. 

**Screenshots**
screen shot on 2018-07-10 at 14_23_15
screen shot on 2018-07-10 at 13_35_50
screen shot on 2018-07-10 at 13_36_09

**Desktop (please complete the following information):**
 - OS: High Sierra
 - Browser Chrome
 - Version 67.0.3396.99

**Additional context**
Gutenberg version 3.2.0

### PR DESCRIPTION
**Pull request by notnownikki**
## Description

Fixes: https://github.com/WordPress/gutenberg/issues/7872

For embeds that work but we don't have a specific block for (i.e. they're not known to Gutenberg) we want to make the content of the frame non-interactive so that whatever links are supplied in the embed's preview don't get clicked and go to different pages inside the iframe. See #7872 for an example of where this can happen.

Why doesn't this use the `Disabled` component in the embed block? Because we want the iframe to be enabled do you can tab through it as normal, but the content inside should be disabled, and the overlay seems to be an easy way of doing that that has no side effects.

## How has this been tested?

Follow the exampled in #7872 and make sure you can't click away from the embedded preview.

## Types of changes
Bug fix

### CODE REVIEW

**Comment by aduth** [comment]
For what reason should we have different treatment of content for known vs. unknown embeds. What is it about a Tweet preview that we should want it to be interactable, but not for one like Giphy? The Giphy case seems like its a misbehaving embed moreso than an issue of how we approach embeds generally. What stops links in known embeds from doing the same, other than the embed provider knowingly correcting this behavior?

**Comment by notnownikki** [comment]
I was looking at this again and thought that perhaps all embed previews should not be interactive. It makes them much easier to select if you can click anywhere on the preview and not have a video start playing, or some text get focus. Do we really want any of them to be interactive in the editor? They are just previews, after all. Perhaps interactive defaults to false?

**Comment by gziolo** [comment]
Moving to 4.2 so we could focus on bigger UI tasks in 4.1.

**Comment by youknowriad** [comment]
Seems blocked somehow. I'm moving out for 4.2 for now.

**Comment by notnownikki** [comment]
If we can live with #7872 then let's close this. Otherwise I'll fix up the conflicts and get this unblocked.

**Comment by notnownikki** [comment]
Actually, looking at https://github.com/WordPress/gutenberg/issues/11885 we problem want to be able to opt blocks in to this. I'll start unblocking it.

File: packages/components/src/sandbox/index.js [hunk 1]
```diff
@@ -154,8 +154,17 @@ class Sandbox extends Component {
...
+				width: 100%;
+				height: 100%;
+				top: 0;
+				left: 0;
+				right: 0;
+				bottom: 0;
+			}
 		`;
-
+		const overlayId = 'wp-sandbox-overlay';
```
Comment (aduth) [inline packages/components/src/sandbox/index.js]: I'm not sure I see a benefit to assigning this to a variable if it's not in-fact a single source of truth (i.e. duplicated in the style assignment).

File: packages/block-library/src/embed/embed-preview.js [hunk 2]
```diff
@@ -45,6 +45,7 @@ const EmbedPreview = ( props ) => {
 				scripts={ scripts }
 				title={ iframeTitle }
 				type={ sandboxClassnames }
+				interactive={ interactive }
```
Comment (aduth) [inline packages/block-library/src/embed/embed-preview.js]: Out of curiosity, is there any reason we couldn't use `Disabled` for this?

https://github.com/WordPress/gutenberg/tree/master/packages/components/src/disabled

**Comment by notnownikki** [comment]
Disabled changes the tab order I think, I wanted to keep it consistent

**Comment by notnownikki** [comment]
Circling back round to this one again, as https://github.com/WordPress/gutenberg/issues/1993#issuecomment-447631339 reminded me how much this trips me up too.

I'll rework it so that the preview iframe is disabled while the block doesn't have focus, so that initial click shouldn't trigger anything, but will select the block. That might get rid of 90% of the issues here.

File: packages/components/src/sandbox/index.js (hunk 1 above)
Comment (notnownikki) [reply packages/components/src/sandbox/index.js]: Me neither, but the linter doesn't allow a string literal as `id` :)

**Comment by notnownikki** [comment]
The approach with `Disabled` causes bugs here, if we disable only when the block is not selected.

1) An optional `Disabled` wrapping element causes the preview to be re-rendered when the block gets selected. This results in significant UI jumps.

2) Clicking on a disabled preview stops mouse events bubbling, and so Gutenberg goes not realise that the mouse button has been released, only that it has been pressed. So when you move your mouse, you start multi-selecting blocks.

I'll try the same approach, but with the overlay instead and see how it goes.

**Comment by notnownikki** [comment]
Tried with the overlay method and it has the same problem as using `Disabled` - you end up multi selecting because the mouseup event doesn't get processed and Gutenberg things you're dragging. I'm not sure why this is yet.

I think the solution might be to have an overlay that stays there, and allows click events to propagate depending on `isSelected`. That will avoid re-rendering the preview iframe and overlay, if it works.

**Comment by notnownikki** [comment]
> I think the solution might be to have an overlay that stays there, and allows click events to propagate depending on isSelected.

And that doesn't work either, because by the time the click event gets to the overlay, the block is selected.

I think it's going to have to be some state in the `EmbedPreview` component that keeps track of if the block has just been selected, and doesn't let that click though, but lets through subsequent clicks.

**Comment by notnownikki** [comment]
Just posting my findings again...

So if you click on an element that is there when the block is not selected, but absent when it is selected (in the case of `Disabled` or an overlay div that serve to stop the initial click making it through to the embedded content) then Gutenberg doesn't pick up on the mouseup event and you're selecting multiple blocks when you move the mouse around. Having a `Disabled` component disappear also resulted in a re-render of the preview, which made UI element jump about the screen while it rendered again. That is also the case if an overlay div was turned on and off inside the sandbox component, because the change in props resulted in a render.

The way to work around all of this is to have an overlay that does not go away at all, but instead we could change the size, or zindex, or something else` to make it appear to disappear. But we can't change `className` because that's a prop change and we would re-render. So, we'd have to get a ref and change the style in the node itself. That'll let the mouse events propagate correctly, and avoid unnecessary renders.

**Comment by notnownikki** [comment]
Closing this in favor of https://github.com/WordPress/gutenberg/pull/12981 which solves it with an overlay and doesn't have the re-rendering or multi-selection issues.

**Pull request closed without merging**